validator/client/service.go

```diff
@@ -227,8 +233,42 @@ func (v *ValidatorService) Start() {
 		eventsChannel:                  make(chan *eventClient.Event, 1),
 	}
 
-	v.validator = valStruct
-	go run(v.ctx, v.validator)
+	hm := newHealthMonitor(v.ctx, v.cancel, v.MaxHealthChecks, v.validator)
+	hm.Start()
+
+	for {
+		select {
+		case <-v.ctx.Done():
+			log.Info("Validator service context canceled, stopping")
+			v.closeClientFunc()
+			return
+		case isHealthy := <-hm.HealthyChan():
+			if !isHealthy {
+				// wait until the next health tracker update
+				log.Warn("Validator service health check failed, waiting for healthy beacon node...")
+				continue
+			}
+			// Reconnect event stream if needed
+			if !v.validator.EventStreamIsRunning() {
+				go v.validator.StartEventStream(v.ctx, eventClient.DefaultEventTopics)
+			}
+
+			log.Info("Starting validator runner")
+			runnerCtx, runnerCancel := context.WithCancel(v.ctx)
```

Comment:
Is it necessary to call `runnerCancel()` before `v.closeClientFunc()`? Because if not, then you can remove calls to `runnerCancel()` and write this:
```suggestion
			runnerCtx, runnerCancel := context.WithCancel(v.ctx)
			defer runnerCancel()
```

Reply:
we should close the entire run if the runner fails on the health loop

Reply:
OK but why call `runnerCancel` twice manually inside this function instead of using `defer`?

Reply:
because we don't exit the loop, the defer i believe triggers at the function level so when Start() completes